### PR TITLE
[16.0][ADD] website_event_membership_restriction: New module

### PR DESCRIPTION
**Pull request by pilarvargas-tecnativa**
This module allows you to restrict event registration to members only.

cc @tecnativa TT49482

@pedrobaeza @carolinafernandez-tecnativa please review

event

The same for logged in users whose associated commercial partner does not have membership_state ‘paid’ or ‘free’.

### CODE REVIEW

**Comment by carolinafernandez-tecnativa** [comment]
Functional review LGTM, Translations are missing.

**Comment by pilarvargas-tecnativa** [comment]
> Functional review LGTM, Translations are missing.

Done! thanks

**Review by carolinafernandez-tecnativa — APPROVED** [review]

File: website_event_membership_restriction/views/event_templates.xml [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        >
+            <attribute name="class">d-none</attribute>
+        </xpath>
+        <xpath
+            expr="//form//div[@id='o_wevent_tickets_collapse']//button"
+            position="after"
+        >
+            <button
+                type="submit"
+                t-att-class="'btn btn-primary o_wait_lazy_js a-submit ' + (event.allow_members_only and (not request.env.user or request.env.user.partner_id.commercial_partner_id.membership_state not in ['paid', 'free']) and 'disabled' or '')"
```
Comment (chienandalu) [inline website_event_membership_restriction/views/event_templates.xml]: - Maybe you should check if the user isn't public better `not request.env.user._is_public()`
- Why checking the commercial_partner_id state? What if membership is given by an associaate?

Comment (pedrobaeza) [reply website_event_membership_restriction/views/event_templates.xml]: @chienandalu do you mean a prorated membership?

File: website_event_membership_restriction/views/event_templates.xml [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            </button>
+            <t
+                t-if="event.allow_members_only and (not request.env.user or request.env.user.partner_id.commercial_partner_id.membership_state not in ['paid', 'free'])"
+            >
+                <a class="text-center small" href="/membership-info">Members only</a>
+            </t>
+        </xpath>
+    </template>
+
+    <template id="website_event_membership_info" name="Membership Info">
```
Comment (pedrobaeza) [inline website_event_membership_restriction/views/event_templates.xml]: This should be a website.page with its `arch`, not a template to be rendered with a controller.

File: website_event_membership_restriction/readme/USAGE.rst [hunk 3]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (pedrobaeza) [inline website_event_membership_restriction/readme/USAGE.rst]: Remove this extra initial line.

File: website_event_membership_restriction/readme/CONFIGURE.rst [hunk 4]
```diff
@@ -0,0 +1,4 @@
+#. Go to *Events > Events*
+#. Create or select one event.
+#. Check or uncheck "Allow registration only to members".
+
```
Comment (pedrobaeza) [inline website_event_membership_restriction/readme/CONFIGURE.rst]: Remove this empty line

File: website_event_membership_restriction/views/event_templates.xml (hunk 1 above)
Comment (chienandalu) [reply website_event_membership_restriction/views/event_templates.xml]: No, I mean that a contact could be assoiciated to other one membership: https://github.com/odoo/odoo/blob/5fa953a0886c0b45169fc894c6202ede2f68f700/addons/membership/models/partner.py#L13-L15

So a commercial partner could have some child contacts which inherit his membership and others that don't. Also contacts inheriting membership from other organization...

Comment (pedrobaeza) [reply website_event_membership_restriction/views/event_templates.xml]: That's not the use case, but let's fallback to check if the commercial partner or the partner itself.

File: website_event_membership_restriction/models/event_event.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# Copyright 2024 Tecnativa - Pilar Vargas
+from odoo import fields, models
+
+
+class EventEvent(models.Model):
+    _inherit = "event.event"
+
+    allow_members_only = fields.Boolean(help="Allow registration only to members")
```
Comment (pedrobaeza) [inline website_event_membership_restriction/models/event_event.py]:
```suggestion
    registration_membership_only = fields.Boolean(help="Allow registration only to members")
```

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: website_event_membership_restriction/models/event_event.py (hunk 5 above)
Comment (pilarvargas-tecnativa) [reply website_event_membership_restriction/models/event_event.py]: All changes done! thanks

File: website_event_membership_restriction/views/event_templates.xml [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        >
+            <attribute name="class">d-none</attribute>
+        </xpath>
+        <xpath
+            expr="//form//div[@id='o_wevent_tickets_collapse']//button"
+            position="after"
+        >
+            <button
+                type="submit"
+                t-att-class="'btn btn-primary o_wait_lazy_js a-submit ' + (event.registration_membership_only and (not request.env.user or request.env.user.partner_id.commercial_partner_id.membership_state not in ['paid', 'free'] or request.env.user.partner_id.membership_state not in ['paid', 'free']) and 'disabled' or '')"
```
Comment (chienandalu) [inline website_event_membership_restriction/views/event_templates.xml]: Do this condition make sense? Should it be something like
```
event.registration_membership_only and not request.env.user._is_public() and (... or ...)

Comment (pilarvargas-tecnativa) [reply website_event_membership_restriction/views/event_templates.xml]: Change made to check with ‘is_public()’.

This way logging will be disabled for the following cases:

- registration_membership_only = true and user not logged in.
- registration_membership_only = true and partner state not free, paid
- registration_membership_only = true and state of partner commercial not free, paid


thanks!

**Review by pedrobaeza — APPROVED** [review]
/ocabot merge nobump

**Pull request merged**